### DISCUSSION TITLE
Maker.js 'toJscadSTL' Function Requires StatusCallback Even When Optional

### DISCUSSION THREAD

**Discussion by ManikandanZST** [microsoft/maker.js#570] [General]
I encountered an issue while using the Maker.js library's to JscadSTL function, which is responsible for exporting 3D models in STL format. According to the documentation, the statusCallback parameter is marked as optional. However, in my experience, I found that the function expects a statusCallback even when it's optional, and not providing one results in an error. Can anyone help me with this @danmarshall ?

**Comment by danmarshall** [comment]
can you paste the specific error?

**Reply by ManikandanZST** [reply]
Thank you for quick response @danmarshall 

Certainly, here is the specific error I've encountered in my Maker.js code:

without statuscallback function:

  protected export(): void {

    let options = {
        stroke: DEFAULT_SVG_COLOR,
        fill: DEFAULT_SVG_COLOR,
        strokeWidth: DEFAULT_STROKE_WIDTH,
        layerOptions: {
            fill: DEFAULT_SVG_COLOR,
        },
    };

    const exported_cad = Builder.makerjs.exporter.toJscadSTL(
     this.cad_object,
     options,
    );

    this.exported_cad = exported_cad;
  }

With statuscallback function:

protected export(statusCallback: (progress: any) => void): void {

     const dummyStatusCallback = (progress: any) => {
       console.log(progress)
   };

     const options = {
       statusCallback: dummyStatusCallback,
  };

    const exported_cad = Builder.makerjs.exporter.toJscadSTL(
        this.cad_object,
        options
    );

    this.exported_cad = exported_cad;
}

Uncaught TypeError: Cannot read properties of undefined (reading 'statusCallback')
    at Object.toJscadSTL (index.js:7004:38)
    at RectangleBoltFlange.export (PartsModel.ts:32:51)
    at RectangleBoltFlange.save (PartsModel.ts:49:10)
    at RectangleBoltFlange.generate (rectangle_bolt_flange.ts:159:10)
    at Rectangle_bolt_flange.jsx:41:69
    at commitHookEffectListMount (react-dom.development.js:23150:26)
    at commitPassiveMountOnFiber (react-dom.development.js:24926:13)
    at commitPassiveMountEffects_complete (react-dom.development.js:24891:9)
    at commitPassiveMountEffects_begin (react-dom.development.js:24878:7)
    at commitPassiveMountEffects (react-dom.development.js:24866:3)
    
It appears that the error is originating from the toJscadSTL function in Maker.js. The error message suggests that it is unable to read properties related to 'statusCallback,' which is causing the issue. Getting the same error for the two code.

I've reviewed the code, and it seems that I am not explicitly providing a statusCallback parameter to the toJscadSTL function in my code. However, this parameter is marked as optional in the Maker.js documentation. Is there a specific way I should be handling this optional parameter to avoid this error?

Any guidance on resolving this issue would be greatly appreciated.

**Reply by ManikandanZST** [reply]
Is it possible to export a Solid 3D object using maker.js?

**Reply by danmarshall** [reply]
Looks like a bug. For now, can you provide a noop function `()=>{}` ?
Yes you can output STL files but you'll need to include some JSCAD libraries. (it may not work with latest JSCAD, but here's a copy of compatible versions: https://github.com/microsoft/maker.js/tree/master/docs/external/jscad)
https://maker.js.org/docs/api/modules/makerjs.exporter.html#tojscadstl

**Reply by ManikandanZST** [reply]
Thank you @danmarshall . It will be helpful to move forward with my work!
